File: generated-code/javascript/codecraft/FileReadWrite/model/debug-interface/debug-data.js
const ColoredVertex = require.main.require('./model/debug-interface/colored-vertex');
const PrimitiveType = require.main.require('./model/debug-interface/primitive-type');

/**
 * Debug data can be drawn in the app
 */
class DebugData {
    /**
     * Read DebugData from input stream
     */
    static async readFrom(stream) {
        let tag = await stream.readInt();
        if (tag == Log.TAG) {
            return await Log.readFrom(stream);
        }
        if (tag == Primitives.TAG) {
            return await Primitives.readFrom(stream);
        }
        if (tag == PlacedText.TAG) {
            return await PlacedText.readFrom(stream);
        }
        throw new Error("Unexpected tag value");
    }
}
/**
 * Log some text
 */
class Log extends DebugData {
    /**
     * Text to show
     */
    text;

    constructor(text) {
        super();
        this.text = text;
    }

    /**
     * Read Log from input stream
     */
    static async readFrom(stream) {
        let text;
        text = await stream.readString();
        return new Log(text);
    }

    /**
     * Write Log to output stream
     */
    async writeTo(stream) {
        await stream.writeInt(Log.TAG);
        let text = this.text;
        await stream.writeString(text);
    }
}

Log.TAG = 0;
DebugData.Log = Log;
/**
 * Draw primitives
 */
class Primitives extends DebugData {
    /**
     * Vertices
     */
    vertices;
    /**
     * Primitive type
     */
    primitiveType;

    constructor(vertices, primitiveType) {
        super();
        this.vertices = vertices;
        this.primitiveType = primitiveType;
    }

    /**
     * Read Primitives from input stream
     */
    static async readFrom(stream) {
        let vertices;
        vertices = [];
        for (let verticesCount = await stream.readInt(); verticesCount > 0; verticesCount--) {
            let verticesElement;
            verticesElement = await ColoredVertex.readFrom(stream);
            vertices.push(verticesElement);
        }
        let primitiveType;
        primitiveType = await PrimitiveType.readFrom(stream);
        return new Primitives(vertices, primitiveType);
    }

    /**
     * Write Primitives to output stream
     */
    async writeTo(stream) {
        await stream.writeInt(Primitives.TAG);
        let vertices = this.vertices;
        await stream.writeInt(vertices.length);
        for (let verticesElement of vertices) {
            await verticesElement.writeTo(stream);
        }
        let primitiveType = this.primitiveType;
        await primitiveType.writeTo(stream);
    }
}

Primitives.TAG = 1;
DebugData.Primitives = Primitives;
/**
 * Draw text
 */
class PlacedText extends DebugData {
    /**
     * Vertex to determine text position and color
     */
    vertex;
    /**
     * Text
     */
    text;
    /**
     * Text alignment (0 means left, 0.5 means center, 1 means right)
     */
    alignment;
    /**
     * Font size in pixels
     */
    size;

    constructor(vertex, text, alignment, size) {
        super();
        this.vertex = vertex;
        this.text = text;
        this.alignment = alignment;
        this.size = size;
    }

    /**
     * Read PlacedText from input stream
     */
    static async readFrom(stream) {
        let vertex;
        vertex = await ColoredVertex.readFrom(stream);
        let text;
        text = await stream.readString();
        let alignment;
        alignment = await stream.readFloat();
        let size;
        size = await stream.readFloat();
        return new PlacedText(vertex, text, alignment, size);
    }

    /**
     * Write PlacedText to output stream
     */
    async writeTo(stream) {
        await stream.writeInt(PlacedText.TAG);
        let vertex = this.vertex;
        await vertex.writeTo(stream);
        let text = this.text;
        await stream.writeString(text);
        let alignment = this.alignment;
        await stream.writeFloat(alignment);
        let size = this.size;
        await stream.writeFloat(size);
    }
}

PlacedText.TAG = 2;
DebugData.PlacedText = PlacedText;
module.exports = DebugData;

File: generated-code/scala/codecraft/FileReadWrite/src/main/scala/trans_gen_test/model/debug_interface/DebugData.scala
package trans_gen_test.model.debug_interface

import trans_gen_test.util.StreamUtil

/**
 * Debug data can be drawn in the app
 */
sealed trait DebugData {
    /**
     * Write DebugData to output stream
     */
    def writeTo(stream: java.io.OutputStream)
}

object DebugData {
    /**
     * Log some text
     *
     * @param text Text to show
     */
    case class Log(text: String) extends DebugData {
        /**
         * Write Log to output stream
         */
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Log.TAG)
            StreamUtil.writeString(stream, text)
        }
    
        /**
         * Get string representation of Log
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Log { ")
            stringBuilder.append("text: ")
            stringBuilder.append('"' + text + '"')
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Log {
        val TAG: Int = 0
    
        /**
         * Read Log from input stream
         */
        def readFrom(stream: java.io.InputStream): Log = Log(
            StreamUtil.readString(stream)
        )
    }

    /**
     * Draw primitives
     *
     * @param vertices Vertices
     * @param primitiveType Primitive type
     */
    case class Primitives(vertices: Seq[trans_gen_test.model.debug_interface.ColoredVertex], primitiveType: trans_gen_test.model.debug_interface.PrimitiveType) extends DebugData {
        /**
         * Write Primitives to output stream
         */
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Primitives.TAG)
            StreamUtil.writeInt(stream, vertices.length)
            vertices.foreach { value =>
                value.writeTo(stream)
            }
            primitiveType.writeTo(stream)
        }
    
        /**
         * Get string representation of Primitives
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Primitives { ")
            stringBuilder.append("vertices: ")
            stringBuilder.append(vertices)
            stringBuilder.append(", ")
            stringBuilder.append("primitiveType: ")
            stringBuilder.append(primitiveType)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Primitives {
        val TAG: Int = 1
    
        /**
         * Read Primitives from input stream
         */
        def readFrom(stream: java.io.InputStream): Primitives = Primitives(
            (0 until StreamUtil.readInt(stream)).map { _ =>
                trans_gen_test.model.debug_interface.ColoredVertex.readFrom(stream)
            },
            trans_gen_test.model.debug_interface.PrimitiveType.readFrom(stream)
        )
    }

    /**
     * Draw text
     *
     * @param vertex Vertex to determine text position and color
     * @param text Text
     * @param alignment Text alignment (0 means left, 0.5 means center, 1 means right)
     * @param size Font size in pixels
     */
    case class PlacedText(vertex: trans_gen_test.model.debug_interface.ColoredVertex, text: String, alignment: Float, size: Float) extends DebugData {
        /**
         * Write PlacedText to output stream
         */
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, PlacedText.TAG)
            vertex.writeTo(stream)
            StreamUtil.writeString(stream, text)
            StreamUtil.writeFloat(stream, alignment)
            StreamUtil.writeFloat(stream, size)
        }
    
        /**
         * Get string representation of PlacedText
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("PlacedText { ")
            stringBuilder.append("vertex: ")
            stringBuilder.append(vertex)
            stringBuilder.append(", ")
            stringBuilder.append("text: ")
            stringBuilder.append('"' + text + '"')
            stringBuilder.append(", ")
            stringBuilder.append("alignment: ")
            stringBuilder.append(alignment)
            stringBuilder.append(", ")
            stringBuilder.append("size: ")
            stringBuilder.append(size)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object PlacedText {
        val TAG: Int = 2
    
        /**
         * Read PlacedText from input stream
         */
        def readFrom(stream: java.io.InputStream): PlacedText = PlacedText(
            trans_gen_test.model.debug_interface.ColoredVertex.readFrom(stream),
            StreamUtil.readString(stream),
            StreamUtil.readFloat(stream),
            StreamUtil.readFloat(stream)
        )
    }

    /**
     * Read DebugData from input stream
     */
    def readFrom(stream: java.io.InputStream): DebugData = {
        StreamUtil.readInt(stream) match {
            case Log.TAG => Log.readFrom(stream)
            case Primitives.TAG => Primitives.readFrom(stream)
            case PlacedText.TAG => PlacedText.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}